Reduce duplicated code in Analytics controllers

tomalec opened the pull request:
_:warning: this PR targets a feature branch together with [multiple improvements to the Analytics codebase](https://github.com/woocommerce/woocommerce/pull/49425)_

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
- [Reuse GenericController::prepare_item_for_response for `Reports\Controller` subclasses](https://github.com/woocommerce/woocommerce/commit/8bdccff32fa6d2d6c375a03ae7248bade75c2eb8) 
- [Reuse GenericController::get_collection_params for `Reports\Controller` subclasses.](https://github.com/woocommerce/woocommerce/commit/7b17eb3d1d8e75c95eb970fdbd44119bfbf86512) 
- [Move get_items code to shared GenericStatsController](https://github.com/woocommerce/woocommerce/commit/a2de7a8f6bbcdd49bca042a509d95eae38331a67)
- [Move shared paginable controlles get_items code to the generic class](https://github.com/woocommerce/woocommerce/pull/49427/commits/2bf60900b516cffa75d4c1bc1b837862b63a7b94)
- [Move fields param to GenericStatsController](https://github.com/woocommerce/woocommerce/pull/49427/commits/28dbf0bbdc2b209109ab3585bef7c8e33926172e)
- [Separate OrderAwareControllerTrait from ReportController](https://github.com/woocommerce/woocommerce/pull/49427/commits/07ed67381782997eb2d67a927340f475ba0f9fea) 
   to allow specific Analytics Controllers extends Generic(Stats)Controller directly, without extending the ReportController which is used to list reports.


Basically, it's meant not to change any behavior. However, due to unification, it did tweak a few things mostly from the developer perspective:
- Previously, only some controllers threw an error from `get_items` when the data was missing; some did not. Now controllers the following Controllers changed behavior to also for such an error:
	- Coupons
	- Customers
	- Downloads
	- Orders
	- Products
	- Taxes
- the error has one error code fro all controllers: `woocommerce_rest_reports_invalid_response` (previously there were report-specific)
   - `woocommerce_rest_reports_categories_invalid_response` -> `woocommerce_rest_reports_invalid_response`
- In Orders Controller the `$orders_data['order_number'], $orders_data['total_formatted']` are set in the `prepare_item_for_response` method not in the `get_items`So `prepare_item_for_response` function is expected to be called with bare data from Query
- In Products Controller extended-info is now sanitized in the `prepare_item_for_response` method, not in the `get_items`
- :warning: (breaking) Previosly some Controlers' `prepare_item_for_response` function was getting and expecting first argument to be casted to `object`, while the rest of controllers used data as returned from the Store. This PR unifies this behavior  for
   -  `Coupons\Stats\Controller`
   -  `Taxes\Controller`
   -  `Taxes\Stats\Controller`
   So, if you extend those classes and override `prepare_item_for_response` in a way that expects `object`, you will get an error. (I have not found such cases in the `all-plugins` repo.)
   All `woocommerce_rest_prepare_report_*` filters remain intact and inconsistent for backward compatibility.


There is another change that I extracted to track separately, as it may introduce more of a breaking change, but on the other hand, may be useful for some long term cleanups
https://github.com/woocommerce/woocommerce/pull/50002

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR aims to change only the underlying code without changing any user-facing functionality, so there are no new things to test, only the regression.

1. Smoke test all Analytics pages, including `WooCommerce > Customers`
2. Check that the code of individual Controllers indeed gets simpler, shorter, and easier to maintain

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Reduce duplicated code in Analytics controllers
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>
…troller` subclasses.

tomalec committed:
Reuse `GenericController::prepare_item_for_response` for `Reports\Controller` subclasses.

tomalec committed:
Reuse `GenericController::get_collection_params` for `Reports\Controller` subclasses.

tomalec committed:
Move get_items code to shared GenericStatsController

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

tomalec committed:
Remove unused `use ParameterException`s

tomalec committed:
Fix default orderby value for Categories controller

tomalec committed:
Move shared paginable controlles get_items code to the generic class

tomalec committed:
Merge branch 'dev/cleanup-controllers' of github.com:woocommerce/woocommerce into dev/cleanup-controllers

tomalec committed:
Move fields param to GenericStatsController

tomalec committed:
Separate  OrderAwareControllerTrait from ReportController

to allow specific Analytics Controllers extends Generic(Stats)Controller directly, without extending the ReportController which is used to list reports.

tomalec committed:
Fix text domain

tomalec committed:
Add whitespace fixes for PHPCS

tomalec committed:
Add since tags for old filters

tomalec committed:
Add whitespace fixes for PHPCS

tomalec commented:
Adding https://github.com/woocommerce/woocommerce/pull/49722 as the changes from there needs to be incorporated here.

tomalec committed:
Unify the data type for `prepare_item_for_response`

to be the one returned by DataStore w/o any casting.
Let `prepare_item_for_response` cast it if needed.

tomalec committed:
Merge branch 'dev/cleanup-analytics' into dev/cleanup-controllers

tomalec committed:
Add `@since` tag to `woocommerce_rest_prepare_report_variations` docs

found origin at https://github.com/woocommerce/woocommerce/commit/164a67144b2182e38dbdca4aec9fc065c7e36a19

tomalec committed:
Misalign spaces for PHPCS

tomalec committed:
and align spaces back for phpcs check

tomalec committed:
Make Dwonloads Controller extend GenericController directyl

leftover from https://github.com/woocommerce/woocommerce/pull/49427/commits/07ed67381782997eb2d67a927340f475ba0f9fea

tomalec committed:
Fix the query class used by Order/Stats/Controller

tomalec committed:
Add whitespaces back and  spli assignemnts with comment

to silence the PHPCS

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/GenericStatsController.php:
```diff
@@ -155,4 +165,60 @@ public function get_item_schema() {
+		);
+	}
+
+	/**
+	 * Maps query arguments from the REST request, to be fed to Query.
+	 *
+	 * `WP_REST_Request` does not expose a method to return all params covering defaults,
+	 * as it does for `$request['param']` accessor.
+	 * Therefore, we re-implement defaults resolution.
+	 *
+	 * @param \WP_REST_Request $request Full request object.
+	 * @return array Simplified array of params.
+	 */
+	protected function prepare_reports_query( $request ) {
+		$args = wp_parse_args(
+			array_intersect_key(
+				$request->get_query_params(),
+				$this->get_collection_params()
+			),
+			$request->get_default_params()
+		);
+
+		return $args;
+	}
```
I haven't started the review but just found something while I was working on an example plugin for the integration guide using this branch. I feel like this method should be defined in `GenericController`, or define a placeholder method for it and throws "not implemented" error. What do you think?

tomalec replied:
I lean towards moving this as is for `GenericConmponent`. This way, new Controllers would not have to re-implement it if they are fine with mapping them from `get_collection_params`.

My only doubt is that this change would affect API of some Controllers that implements their own `get_items` and  doesn't use or have `prepare_reports_query` method:
- `plugins/woocommerce/src/Admin/API/Reports/Controller.php`
- `plugins/woocommerce/src/Admin/API/Reports/Export/Controller.php`
- `plugins/woocommerce/src/Admin/API/Reports/Import/Controller.php`

And due to out long-term backward compatibility policy, I'd be cautious not to add too much.

WDYT?

ianlin replied:
Sorry if I was being unclear, I guess the "throwing not implemented error" idea caused confusion and it isn't the best approach for backward compatibility.

I meant that currently, the prepare_reports_query() method is added to the `GenericStatsController` rather than the `GenericController`. In my case, I created a controller extending `GenericController` without implementing my own `get_items()` method. When I made a request to my controller, it caused a fatal error: `Call to undefined method prepare_reports_query()`.

Since most of the current controllers under `src/Admin/API/Reports/*`, including both stats and non-stats, do not implement their own `get_items()` method, and that the `GenericStatsController` also extends `GenericController`, I think maybe we could move the `prepare_reports_query()` method from `GenericStatsController` to `GenericController`. I might be missing something as I am not fully familiar with this code base, which has many dependencies.

tomalec replied:
Thanks for the clarification. Yes, now I can see the point. Having a Controller that inherits from `GenericController` on the premise of using defaults and throws errors out of the blue is definitely not the experience we want. In such case having a useless method in those 3 classes is acceptable price to pay.

Fixed in https://github.com/woocommerce/woocommerce/pull/49427/commits/c0edda1964ab85f73c6a426e7be90024a11eb14a

tomalec committed:
Update PHPDoc

tomalec committed:
Move `prepare_reports_query` to `GenericController`

Address https://github.com/woocommerce/woocommerce/pull/49427#discussion_r1694929797

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Categories/Controller.php:
```diff
@@ -189,59 +189,16 @@ public function get_item_schema() {
-			),
-			'validate_callback' => 'rest_validate_request_arg',
+		$params                    = GenericController::get_collection_params();
+		$params['orderby']['enum'] = array(
```
Should we add `$params['orderby']['default']` as `category_id`? Since the default in GenericController is `date`.

ianlin replied:
Saw you fixed it already in 0d12825.

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Controller.php:
```diff
@@ -243,54 +243,11 @@ public function get_item_schema() {
-			'description'       => __( 'Maximum number of items to be returned in result set.', 'woocommerce' ),
-			'type'              => 'integer',
-			'default'           => 10,
-			'minimum'           => 0,
```
The `minimum` for `Orders/Controller` is set to `0` while in GenericController is set to `1`, I'm not sure if `Orders/Controller` has a special case for minimal of `0`, just pointing this out since I noticed it.

tomalec replied:
Thanks for catching it. I don't see a reason of this special case, but I'll change it back for backward compatibility
https://github.com/woocommerce/woocommerce/pull/49427/commits/a9d06bf0c061468ce9c81db321ad59f1f8aad95a

ianlin reviewed (approved):
Thanks for the clean up on the analytics controller, the code looks great and the functionality remains unchanged. I also test it using an [example plugin](https://github.com/woocommerce/current-wc-analytics-examples) and it works as expected.

tomalec committed:
Bring back minimum value of 0 for per_page to Orders Controller

Address https://github.com/woocommerce/woocommerce/pull/49427/files/8bdccff32fa6d2d6c375a03ae7248bade75c2eb8..7b17eb3d1d8e75c95eb970fdbd44119bfbf86512#r1706910760

tomalec commented:
Some E2Es are failing in CI 
![image](https://github.com/user-attachments/assets/6a9cb1b2-3a6b-456f-bfb5-dcf9389f45fe)

However, locally, I'm able to click and install Mailpoet. Maybe merging from the trunk will help. I'll merge this PR to the feature branch so we can work on tests there, reducing the number of merges back and forth.
